### DISCUSSION TITLE
8주차: 11장 API 리팩터링

### DISCUSSION THREAD

**Discussion by choisohyun** [frontend-book-study/refactoring-2#18] [리팩터링 2판]
|							| 질문			| 답변			|
|-------------------------------------------------------|----------------------|----------------------|
| 1		|  @LastCow9000		| @choisohyun		|
| 2		| @choisohyun		|  @hamo-o		|
| 3		|  @hamo-o		| @LastCow9000		|

**Comment by choisohyun** [comment]
- 다음 두 가지 리팩토링(매개변수를 질의 함수로 바꾸기 / 질의 함수를 매개변수로 바꾸기)을 어떤 상황일 때 하면 좋을까요?

**Reply by hamo-o** [reply]
**매개변수를 질의 함수로 바꾸기**

🙆‍♀️ 해야 할 때 🙆‍♀️
매개변수 목록에서의 중복을 피해야 할 때
대상 함수가 스스로 “쉽게” 구할 수 있는 값을 매개변수로 받고 있을 때

🙅‍♀️ 하지 말아야 할 때 🙅‍♀️
매개변수 제거 시 대상 함수에 원치 않는 의존성이 생길 때
대상 함수가 참조 투명(함수에 똑같은 값을 주어서 호출하면 항상 똑같이 동작)하지 않을 때

---

**질의 함수를 매개변수로 바꾸기**

🙆‍♀️ 해야 할 때 🙆‍♀️
대상 함수가 더 이상 특정 원소에 의존하길 원하지 않을 때 (전역 변수, 제거해야 하는 원소의 참조, ...)
대상 함수가 참조 투명성을 잃었다면, 문제를 유발하는 원소를 매개변수로 바꿔 해결할 수 있다.

🙅‍♀️ 하지 말아야 할 때 🙅‍♀️
호출하는 함수가 어떤 값을 넘겨줘야 할 지 계산하는 책임이 너무 복잡해질 때

---

매개변수를 질의 함수로 바꾸기(피호출자 책임 크다) ↔️ 질의 함수를 매개변수로 바꾸기(호출자 책임 크다)

**Comment by LastCow9000** [comment]
## 1번 질문
Q1. 참조 투명성이란 무엇이고 참조 투명성을 지키는 함수는 무엇이 좋을까요?
Q2. 질의 함수와 변경 함수를 분리(command-query-separation)하면 얻는 이점은 무엇이 있을까요?

**Reply by choisohyun** [reply]
A1. 똑같은 값을 건네면 매번 똑같은 결과를 내는 함수는 다루기 쉽고 이런 성질을 참조 투명성이라고 합니다. 참조가 투명하지 않은 원소를 매개변수로 바꾸면 해결이 됩니다.

**Reply by choisohyun** [reply]
A2. 겉보기 부수효과가 없어 어떤 때에 호출해도 같은 값을 반환합니다. 겉보기 부수효과는 요청된 값을 캐시해두고 다음번 호출 대 빠르게 응답하는 방법이 있어 이런 캐싱도 객체의 상태를 변경하기 때문입니다.

**Comment by hamo-o** [comment]
Q1. 명령-질의 분리 규칙의 내용을 설명해주세요. (11.1)
Q2. "함수를 명령으로 바꾸기"에서 언급된 "명령" 이란 무엇일까요? 또 이 "명령"은 앞선 "명령-질의 분리" 에서의 명령과 같을까요? (11.9)

**Reply by LastCow9000** [reply]
A1. 명령은 CRUD에서 CUD. 즉, 생성, 수정, 삭제에 해당하고 질의는 CRUD에서 R, 즉, 조회에 해당합니다. 조회 함수에서는 읽기 작업만 하여 몇 번을 조회하든 멱등성을 보장하여 사이드 이펙트를 발생시키지 않는 장점을 가지고 있습니다.

A2. 다릅니다. "함수를 명령으로 바꾸기"에서 언급된 "명령" 은 디자인 패턴 중 `command 패턴에서 사용하는 명령 인 실행될 기능을 캡슐화한 것`에 해당하고 "명령-질의 분리" 에서의 명령은 `값을 수정하는 것`을 의미합니다.
